man/rtgpoll.1
.TH rtgpoll 1 "August 2003" "Manual page for rtgpoll"
.SH NAME
.I rtgpoll
\- RTG SNMP polling daemon
.SH SYNOPSIS
.B rtgpoll
\-t file [options]
.br
.SH DESCRIPTION
.I rtgpoll
queries SNMP-capable devices and inserts the results of these 
queries into a MySQL database.  The poller reads a fixed-format
target file, randomizes the targets and maintains a set number
of queries in flight.  
.I rtgpoll
optionally reads a configuration file to modify run-time
parameters.
.SH OPTIONS
.PP
.TP
.IR "\-t file"
Target file.  See target file section below for additional details.
.TP
.IR "\-c file"
Configuration file.  See configuration file section below for additional 
details.
.TP
.IR "\-d"
Disable database inserts.
.TP
.IR "\-m"
Skip checking for multiple instances.
.TP
.IR "\-z"
Force database insertion of zero deltas.
.TP
.IR "\-v"
Increase verbosity by one level.  Four verbosity levels exist 
(in increasing order): Off, Low, High, Debug.
.SH "CONFIGURATION FILE"
.PP
.br
All RTG programs attempt to find an rtg.conf file in the following
path order:
.PP
1. ./rtg.conf
.br
2. /usr/local/rtg/etc/rtg.conf
.br
3. /etc/rtg.conf
.PP
If rtgpoll or rtgplot cannot find an rtg.conf file in any of these 
paths, it will attempt to create one in the current working directory.
Most users maintain a master rtg.conf file customized for their 
environment.  rtg.conf contains the following configurable fields:
.PP
  Interval         300
.br
  HighSkewSlop     3.0
.br
  LowSkewSlop      0.5
.br
  OutOfRange       93750000000
.br
  SNMP_Ver         1
.br
  SNMP_Port        161
.br
  DB_Host          localhost
.br
  DB_Database      rtg
.br
  DB_User          snmp
.br
  DB_Pass          rtgdefault
.br
  Threads          5
.PP
Interval is the time between successive polls of the target list, default
is 300 seconds (5 minutes).  HighSkewSlop defines the maximum number of
Intervals allowed between two consecutive poll values before the time in
seconds between said points is deemed too large to calculate a valid rate.
With the default Interval and HighSkewSlop values, that time would be 300
* 3 (15 minutes).  LowSkewSlop defines the minimum number of Intervals
alloweed between two consecutive poll values before the time in seconds
between said points is deemed two low to calculate a valid rate.  With the
default Interval and LowSkewSlop values, that time would be 300 * 0.5 (2.5
minutes).  OutOfRange defines an upper bound above which rtgpoll will never
attempt an insert into the database. OutOfRange should be a multiple of
the maximum number of bytes possible in the defined Interval for your
highest speed link.  The default OutOfRange value will suffice in most
installations.  SNMP_Ver specifies the SNMP version the poller will use.  
The number of threads rtgpoll will use is defined in the variable Threads.

Variables in rtg.conf must match the names above exactly.  Comments
and blank lines are allowed and the ordering of variables in rtg.conf
does not matter.
.PP
.SH "TARGET FILE"
.PP
The target file specifies the objects to be SNMP polled.  Comments must be
preceded with a '#' sign.  Elements in the target file are tab delimited.
The format of the target file is fixed:
.PP
  # Host  OID  64/32  Community  Table   ID   Description
.PP
where
.PP
  Host        = IP or hostname of target
.br
  OID         = Full SNMP OID, e.g. .1.3.6.1.2.1.31.1.1.1.10.19
.br
  64/32/0     = Specify 64/32 bit objects or 0 for gauge objects 
.br
  Community   = SNMP Community
.br
  Table       = MySQL table in the database to use
.br
  ID          = A unique ID that is used with each insert
.br
  Description = Text
.PP
RTG can monitor OIDs that return a gauge value allowing one to monitor
items such as temperature, CPU, users, etc.  If an entry in the target
file has 0 (zero) specified as the OID bit width instead of 64 or 32,
the poller assumes that it is monitoring a gauge and does not attempt
to calculate an interval delta value.
.PP
rtgpoll first reads the configuration file, then the target file.  For
each SNMP poll, rtgpoll will attempt an SQL INSERT of the form:
.PP
  INSERT INTO Table VALUES (ID, NOW(), bigint)
.PP
Where Table is the name of the database table and ID is an integer.  Both
Table and ID come from the target list, NOW() is the current timestamp and
bigint is the delta value between successive SNMP polls.
.PP
.SH "SIGNALS"
.PP
rtgpoll accepts a number of signals.  SIGHUP forces a reload of the target
file.  If rtgpoll is actively polling, the target file is reloaded when
rtgpoll becomes idle.  This is useful when automating the target list
creation based on your active network.  SIGUSR1 increases the verbosity of
a running rtgpoll; SIGUSR2 decreases the verbosity.
.PP
.SH "SEE ALSO"
rtgplot(1)
.br
.SH VERSION
This manual page documents rtgpoll version 0.7.4
.SH AUTHOR
.br
(c) 2002-2003 by Robert Beverly
.PP


NEWS
rtg-0.7.4 (2003-10-02)
---------
* Added fast target hashing in rtgpoll.
* Added plotting of interface aggregates into rtgplot.
* Added multiple simultaneous interface plotting into rtgplot.
* Added y-scaling by interface speed support into rtgplot.  
* Added arbitrary image and border sizing into rtgplot.
* Added using autoconf defined --prefix as search path for rtg.conf.
* Added graph fill of area under plot lines into rtgplot.
* Added rtgpoll option (-m) to skip checking for multiple instances.
* Added rtgpoll option (-z) to force database insertion for zero deltas.
* Added debug file pointer; allows writing debug info to file while 
   running rtgplot via CGI.
* Added contrib directory to top-level distribution tree.
* Added libz probing in configure; cleaned up probing logic.
* Changed autoconf SNMP library probing order to detect Net-SNMP first.
* Changed autoconf and code to use strtoll() if possible.
* Audit code for potential buffer overflows; modified all code to use 
   strncpy(), snprintf, and sscan with width specifier.
* Modified all php to use full tags instead of short tags.
* Thread code cleaned and logic improved.
* Fixed bug in rtgplot that incorrectly counted skewed measurements
   in rate max and average.
* From: Chris Brotsos
   Support for per-target maximum polled values.
* From: Chris Brotsos
   Patch to allow ASN_INTEGER/ASN_OPAQUE type responses. 
* From: Shayne Smyth
   Patch to allow ASN_TIMETICKS type responses.
* From: Chris Brotsos
   Patch to implement a PID file and additional signal catching.
* From: Joost Lek
   Support for Nth percentile lines in rtgplot.
* From: Chris Brotsos
   Patch to allow rtgplot to determine last database value when last
   delta occurs outside of queried range.
* From: Michal Szczotka
   Patch to fix division by zero in 95.pl and report.pl.
* From: Brian T. O'Neill
   Patch for extending YPLOT_AREA with GAUGE data lines to allow visibility. 
* From: Brian T. O'Neill
   Patch for allowing configurable HighSkewSlop and LowSkewSlop (See README).
* From: Brian T. O'Neill
   Patch to implement factor logic with GAUGE data.
* From: Brian T. O'Neill
   Modularized rtgtargmkr included in contrib.
* From: Brian T. O'Neill
   Patch to properly read large interface speeds from database.


rtg-0.7.3 (2003-02-08)
---------
* Fixed bug in poller() where last_value object was improperly updated
   when polled device did not respond.
* Fixed rtgplot.cgi to support apache2.
* Fixed read_rtg_config() to handle comments and arbitrary ordering.
* Fixed NULL pointer crash in rtgsnmp.c snmp_free_pdu().
* Fixed rtgplot to detect, compensate and properly plot missed polls.
* Fixed rtgplot to time-align samples between multiple data sources
   when scalex is defined and the data (in database) for one begins
   before the other.
* Fixed FreeBSD autoconf support.
* Added gauge (absolute SNMP counter) support.
* Added config directive to specify SNMP version (1 or 2).
* Added config directive to specify number of threads.
* Added timestamp to poll begin/complete messages.
* Added poll round counter to poll complete messages.
* Added shading to rtgplot.c
* Added red arrow showing direction of growth to rtgplot graphs.
* Added autoconf detection for libz if required by MySQL.
* Added 95.php, 95th percentile report PHP web script.
* Added view.php, an MRTG-style day/week/month display for RTG in PHP.
* Modified all .pl and .php scripts to find and read from the 
   master rtg.conf first, then fall back to defaults.
* Modified rtgtargmkr.pl to read counter bit size from routers
   configuration file to allow size to be specified per router.
* Modified sleepy() routine to use usleep calls making the sleep
   interval exact and the period between polls accurate.
* Modified rtgtargmkr.pl to write out IPs instead of hostnames so
   that DNS failures do not impact collection.
* Aligned rate text in rtgplot.c properly.
* Updated Simon Leinen's included SNMP_Session to version 0.94 (used 
   by rtgtargmkr.pl).
* From: Jon Dustin
   Patch to rtgtargmkr.pl to perform case insensitive checking of
   reserved interfaces.
* From: Stephen Fromm
   Patch to configure to deal with multiple MySQL installs.


rtg-0.7.2 (2002-10-04)
---------
* Fixed report.pl to handle interfaces with zero speed gracefully.
* Rewrote autoconf to detect either UCD-SNMP or Net-SNMP.
* Fixed incompatibility with ucd/net-snmp versions >= 5.x.  Poller now 
   fixed to use only 5.x SNMP functions.
* Fixed bad SQL update statement in rtgtargmkr.pl
* Made rtgtargmkr.pl Cisco Catalyst aware.
* Added time range argument to 95th percentile report
* From: David A. Snodgrass
   Patch to rtgtargmkr.pl to handle interface descriptions with quotes.


rtg-0.7.1 (2002-08-28)
---------
* Fixed bad overflow with 64-bit SNMP counters.
* Fixed database schema to accommodate long interface names.
* Fixed rtgtargmkr.pl router file parsing.
* Fixed rtgtargmkr.pl bug handling descriptions with trailing whitespace.
* From: Adam Rothschild
   Patch to report.pl to specify time periods.
* From: David A. Snodgrass
   Changed createdb MySQL permissions to be more restrictive.
* From: Omachonu Ogali
   Patch to rtgtargmkr.pl to support per-device communities.


rtg-0.7 (2002-08-07)
-------
* Major rewrite of poller to be entirely multi-threaded.  Assumes
   full pthreads support and MySQL thread-safe client support.  
* Autoconf fixes to support FreeBSD -pthreads, Solaris thread
   compiler flags, checks for long longs.
* Changed insert date to NOW() function to reflect SQL server's current
   time.  This removes any synchronization issues in the RTG pollers.
* Fixed SIGHUP initiated memory leak.
* Cleaned up installation and installation scripts.
* Added Perl DBI reports.
* Added Perl target file maintainer script.
* From: David A. Snodgrass
   Patch to createdb to include autoconf MySQL path


rtg-0.6 (2002-07-23)
-------
* Added randomization of target list to "even out" SNMP query load on 
   network devices.

src/rtgpoll.c
/****************************************************************************
   Program:     $Id: rtgpoll.c,v 1.22 2003/09/25 15:56:04 rbeverly Exp $
   Author:      $Author: rbeverly $
   Date:        $Date: 2003/09/25 15:56:04 $
   Description: RTG SNMP get dumps to MySQL database
****************************************************************************/

#define _REENTRANT
#include "common.h"
#include "rtg.h"

#include <syslog.h>
#include <ganglia.h>

/* Yes.  Globals. */
stats_t stats =
{PTHREAD_MUTEX_INITIALIZER, 0, 0, 0, 0, 0, 0, 0, 0, 0.0};
char *target_file = NULL;
target_t *current = NULL;
int entries = 0;
/* dfp is a debug file pointer.  Points to stderr unless debug=level is set */
FILE *dfp = NULL;
const char *program = "rtgpoll";

Ganglia_pool global_context;
Ganglia_metric gmetric;
Ganglia_gmond_config gmond_config;
Ganglia_udp_send_channels send_channels;

/* Main rtgpoll */
int main(int argc, char *argv[]) {
    crew_t crew;
    pthread_t sig_thread;
    sigset_t signal_set;
    struct timeval now;
    double begin_time, end_time, sleep_time;
    char *conf_file = NULL;
    char *gmond_conf = NULL;
    char errstr[BUFSIZE];
    int ch, i;

	dfp = stderr;

    /* Check argument count */
    if (argc < 3)
	usage(argv[0]);

	/* Set default environment */
    config_defaults(&set);

    /* Parse the command-line. */
    while ((ch = getopt(argc, argv, "c:g:Ghmt:vz")) != EOF)
	switch ((char) ch) {
	case 'c':
	    conf_file = optarg;
	    break;
	case 'g':
	    gmond_conf = optarg;
	    break;
	case 'G':
	    set.goff = TRUE;
	    break;
	case 'h':
	    usage(argv[0]);
	    break;
	case 'm':
	    set.multiple++;
	    break;
	case 't':
	    target_file = optarg;
	    break;
	case 'v':
	    set.verbose++;
	    break;
	case 'z':
	    set.withzeros = TRUE;
	    break;
	}

   //fork unless debug mode
   if (set.verbose == OFF) {
       if (fork() == 0) {
           umask(0);
           setsid();
           close(0);
           close(1);
           close(2);
           if (fork())
               return (0);
       } else {
           return (0);
       }
   }

    /* Initialize syslog */
    openlog(program, LOG_PID, LOG_USER);

    if (set.verbose >= LOW)
	printf("RTG version %s starting.\n", VERSION);
    syslog(LOG_NOTICE, "RTG version %s starting.", VERSION);

    /* Initialize signal handler */
    sigemptyset(&signal_set);
    sigaddset(&signal_set, SIGHUP);
    sigaddset(&signal_set, SIGUSR1);
    sigaddset(&signal_set, SIGUSR2);
    sigaddset(&signal_set, SIGTERM);
    sigaddset(&signal_set, SIGINT);
    sigaddset(&signal_set, SIGQUIT);
	if (!set.multiple) 
    	checkPID(PIDFILE);

    if (pthread_sigmask(SIG_BLOCK, &signal_set, NULL) != 0) {
	printf("pthread_sigmask error\n");
        syslog(LOG_ERR, "pthread_sigmask error!");
    }

    /* Read configuration file to establish local environment */
    if (conf_file) {
      if ((read_rtg_config(conf_file, &set)) < 0) {
         printf("Could not read config file: %s\n", conf_file);
         syslog(LOG_ERR, "Could not read config file: %s", conf_file);
	 closelog();
         exit(-1);
      }
    } else {
      conf_file = malloc(BUFSIZE);
      if (!conf_file) {
         printf("Fatal malloc error!\n");
         syslog(LOG_ERR, "Fatal malloc error!");
         closelog();
         exit(-1);
      }
      for(i=0;i<CONFIG_PATHS;i++) {
        snprintf(conf_file, BUFSIZE, "%s%s", config_paths[i], DEFAULT_CONF_FILE); 
        if (read_rtg_config(conf_file, &set) >= 0) {
           break;
        } 
        if (i == CONFIG_PATHS-1) {
           snprintf(conf_file, BUFSIZE, "%s%s", config_paths[0], DEFAULT_CONF_FILE); 
	   if ((write_rtg_config(conf_file, &set)) < 0) {
	      fprintf(stderr, "Couldn't write config file.\n");
	      syslog(LOG_ERR, "Couldn't write config file '%s'.", conf_file);
              closelog();
	      exit(-1);
	    }
        }
      }
    }
    if (gmond_conf) {
    /* Open the Ganglia config file */
    FILE *gfp;

        if ((gfp = fopen(gmond_conf, "r")) == NULL) {
                fprintf(stderr, "\nCould not open target file for reading '%s'.\n", gmond_conf);
                syslog(LOG_ERR, "Could not open target file for reading '%s'.", gmond_conf);
                closelog();
                exit(1);
        }
    } else {
      gmond_conf = malloc(BUFSIZE);
      if (!gmond_conf) {
         printf("Fatal malloc error!\n");
         syslog(LOG_ERR, "Fatal malloc error!");
         closelog();
         exit(1);
      }
        snprintf(gmond_conf, BUFSIZE, "%s", DEFAULT_GMOND_CONF); 
    }
    if (set.verbose >= LOW)
            printf("\nUsing Ganglia config file [%s].\n", gmond_conf);
    syslog(LOG_NOTICE, "Using Ganglia config file [%s].", gmond_conf);

    /* hash list of targets to be polled */
    if (target_file != NULL) 
	entries = hash_target_file(target_file);
    else {
        fprintf(stderr, "No target file defined. Exiting.\n");
        syslog(LOG_ERR, "No target file defined. Exiting.");
        closelog();
        exit(-1);
    }
    if (entries <= 0) {
	fprintf(stderr, "Error updating target list.\n");
	syslog(LOG_ERR, "Error updating target list.");
	closelog();
	exit(-1);
    }
    if (set.verbose >= LOW)
	printf("Initializing threads (%d).\n", set.threads);
    syslog(LOG_NOTICE, "Initializing threads (%d).", set.threads);
    pthread_mutex_init(&(crew.mutex), NULL);
    pthread_cond_init(&(crew.done), NULL);
    pthread_cond_init(&(crew.go), NULL);
    crew.work_count = 0;

    /* Initialize the SNMP session */
    if (set.verbose >= LOW)
	printf("Initializing SNMP (v%d, port %d).\n", set.snmp_ver, set.snmp_port);
    syslog(LOG_NOTICE, "Initializing SNMP (v%d, port %d).", set.snmp_ver, set.snmp_port);
    init_snmp("RTG");

    /* Ganglia setup */
    if (set.verbose >= LOW)
	printf("Initializing Ganglia UDP send channels.\n");
    syslog(LOG_NOTICE, "Initializing Ganglia UDP send channels.");
    /* create the global context */
    global_context = Ganglia_pool_create(NULL);
    if (!global_context) {
        fprintf(stderr, "Unable to create global context. Exiting.\n");
        syslog(LOG_ERR, "Unable to create global context. Exiting.");
        closelog();
        exit(1);
    }

    /* parse the configuration file */
    gmond_config = Ganglia_gmond_config_create(gmond_conf, 0);       // FIXME -- use SYSCONFDIR

    /* build the udp send channels */
    send_channels = Ganglia_udp_send_channels_create(global_context, gmond_config);
    if (!send_channels) {
        fprintf(stderr, "Unable to create ganglia send channels. Exiting.\n");
        syslog(LOG_ERR, "Unable to create ganglia send channels. Exiting.");
        closelog();
        exit(1);
    }
    if (set.verbose >= LOW)
	printf("Ganglia Ready.\n");
    syslog(LOG_NOTICE, "Ganglia Ready.");

    if (set.verbose >= HIGH)
	printf("Starting threads.\n");
    syslog(LOG_NOTICE, "Starting threads.");

    for (i = 0; i < set.threads; i++) {
	crew.member[i].index = i;
	crew.member[i].crew = &crew;
	if (pthread_create(&(crew.member[i].thread), NULL, poller, (void *) &(crew.member[i])) != 0) {
	    printf("pthread_create error\n");
            syslog(LOG_ERR, "pthread_create error");
        }
    }
    if (pthread_create(&sig_thread, NULL, sig_handler, (void *) &(signal_set)) != 0) {
	printf("pthread_create error\n");
	syslog(LOG_ERR, "pthread_create error");
    }

    /* give threads time to start up */
    sleep(1);

    if (set.verbose >= LOW)
	printf("RTG Ready.\n");
    syslog(LOG_NOTICE, "RTG Ready.");

    /* Loop Forever Polling Target List */
    while (1) {
	lock = TRUE;
	gettimeofday(&now, NULL);
	begin_time = (double) now.tv_usec / 1000000 + now.tv_sec;

	PT_MUTEX_LOCK(&(crew.mutex));
	init_hash_walk();
	current = getNext();
	crew.work_count = entries;
	PT_MUTEX_UNLOCK(&(crew.mutex));
	    
	if (set.verbose >= LOW)
        timestamp("Queue ready, broadcasting thread go condition.");
	PT_COND_BROAD(&(crew.go));
	PT_MUTEX_LOCK(&(crew.mutex));
	    
	while (crew.work_count > 0) {
		PT_COND_WAIT(&(crew.done), &(crew.mutex));
	}
	PT_MUTEX_UNLOCK(&(crew.mutex));

	gettimeofday(&now, NULL);
	lock = FALSE;
	end_time = (double) now.tv_usec / 1000000 + now.tv_sec;
	stats.poll_time = end_time - begin_time;
        stats.round++;
	sleep_time = set.interval - stats.poll_time;

	if (waiting) {
	    if (set.verbose >= HIGH)
		printf("Processing pending SIGHUP.\n");
            syslog(LOG_NOTICE, "Processing pending SIGHUP.");
	    entries = hash_target_file(target_file);
	    waiting = FALSE;
	}
	if (set.verbose >= LOW) {
        snprintf(errstr, sizeof(errstr), "Poll round %d complete.", stats.round);
        timestamp(errstr);
        syslog(LOG_NOTICE, "%s", errstr);
	    print_stats(stats);
    }
	if (sleep_time <= 0)
	    stats.slow++;
	else
	    sleepy(sleep_time);

    } /* while */

    exit(0);
}


/* Signal Handler.  USR1 increases verbosity, USR2 decreases verbosity. 
   HUP re-reads target list */
void *sig_handler(void *arg)
{
    sigset_t *signal_set = (sigset_t *) arg;
    int sig_number;

    while (1) {
	sigwait(signal_set, &sig_number);
	switch (sig_number) {
            case SIGHUP:
                if(lock) {
                    waiting = TRUE;
                }
                else {
                    entries = hash_target_file(target_file);
                    waiting = FALSE;
                }
                break;
            case SIGUSR1:
                set.verbose++;
                break;
            case SIGUSR2:
                set.verbose--;
                break;
            case SIGTERM:
            case SIGINT:
            case SIGQUIT:
                if (set.verbose >= LOW)
                   printf("Quiting: received signal %d.\n", sig_number);
                syslog(LOG_WARNING, "Quiting: received signal %d", sig_number);
                unlink(PIDFILE);
                closelog();
                exit(1);
                break;
        }
   }
}


void usage(char *prog)
{
    printf("rtgpoll - RTG v%s\n", VERSION);
    printf("Usage: %s [-Gmz] [-vvv] [-c <rtg.conf>] [-g <gmond.conf>] -t <target.conf>\n", prog);
    printf("\nOptions:\n");
    printf("  -c <file>   Specify RTG configuration file\n");
    printf("  -g <file>   Specify Ganglia agent config file (default: /etc/ganglia/gmond.conf)\n");
    printf("  -G          Disable Ganglia metric reporting\n");
    printf("  -t <file>   Specify SNMP target file\n");
    printf("  -v          Increase verbosity\n");
	printf("  -m          Allow multiple instances\n");
	printf("  -z          Database zero delta inserts\n");
    printf("  -h          Help\n");
    exit(-1);
}
